include/semver.hpp

```diff
@@ -426,16 +458,316 @@ inline std::basic_ostream<Char, Traits>& operator<<(std::basic_ostream<Char, Tra
   return os;
 }
 
-namespace ranges {
+class range {
+ public:
+  constexpr explicit range(std::string_view str) : str_{str} {}
+
+  constexpr bool satisfies(const version& ver) const {
+    return satisfies(ver, false);
+  }
+
+  constexpr bool satisfies_include_prerelease(const version& ver) const {
+    return satisfies(ver, true);
+  }
+
+private:
+  constexpr bool satisfies(const version& ver, bool include_prerelease) const {
+    range_parser parser{str_};
+
+    auto is_logical_or = [&parser]() constexpr -> bool {
+      return parser.current_token.type == range_token_type::logical_or;
+    };
+
+    auto is_operator = [&parser]() constexpr -> bool {
+      return parser.current_token.type == range_token_type::range_operator;
+    };
+
+    auto is_number = [&parser]() constexpr -> bool {
+      return parser.current_token.type == range_token_type::number;
+    };
+
+    const auto has_prerelease = ver.prerelease_type != prerelease::none;
+
+    do
+    {
+      if (is_logical_or()) {
+        parser.advance_token(range_token_type::logical_or);
+      }
+
+      auto contains = true;
+      auto allow_compare = include_prerelease;
+
+      while (is_operator() || is_number()) {
+        const auto range = parser.parse_range();
+        const auto equal_without_tags = (version{range.ver.major, range.ver.minor, range.ver.patch} == version{ver.major, ver.minor, ver.patch});
+
+        if (has_prerelease && equal_without_tags) {
+          allow_compare = true;
+        }
+
+        if (!range.satisfies(ver)) {
+          contains = false;
+          break;
+        }
+      }
+
+      if (has_prerelease) {
+        if (allow_compare && contains) {
+          return true;
+        }
+      } else if (contains) {
+        return true;
+      }
+
+    } while (is_logical_or());
+    
+    return false;
+  }
+
+  enum struct range_operator : std::uint8_t {
+    less,
+    less_or_equal,
+    greater,
+    greater_or_equal,
+    equal
+  };
+
+  struct range_comparator {
+    range_operator op;
+    version ver;
+
+    constexpr bool satisfies(const version& version) const {
+      switch (op) {
+        case range_operator::equal:
+          return version == ver;
+        case range_operator::greater:
+          return version > ver;
+        case range_operator::greater_or_equal:
+          return version >= ver;
+        case range_operator::less:
+          return version < ver;
+        case range_operator::less_or_equal:
+          return version <= ver;
+      }
+      NEARGYE_NODEFAULT
```

Comment:
maebe NEARGYE_THROW(std::invalid_argument{"semver::range unexpected operator."});?
and delete NEARGYE_NODEFAULT